File: backend/src/services/projector.service.js
import ProjectorSession from '../models/ProjectorSession.js';
import Device from '../models/Device.js';
import logger from '../config/logger.js';

/**
 * Create a new projector session
 * @param {Object} sessionData - Session data
 * @returns {Object} Created session
 */
export const createSession = async (sessionData) => {
  try {
    const sessionId = await ProjectorSession.generateSessionId();
    
    const session = await ProjectorSession.create({
      ...sessionData,
      sessionId
    });

    logger.info(`Projector session created: ${sessionId} by ${sessionData.startedBy}`);
    return session;
  } catch (error) {
    logger.error('Create projector session error:', error);
    throw error;
  }
};

/**
 * Get active session for device
 * @param {string} deviceId - Device ID
 * @returns {Object|null} Active session
 */
export const getActiveSession = async (deviceId) => {
  try {
    const session = await ProjectorSession.findOne({
      deviceId,
      status: 'active'
    })
      .populate('classId', 'grade section classCode')
      .populate('startedBy', 'name email')
      .populate('institutionId', 'name');

    return session;
  } catch (error) {
    logger.error('Get active session error:', error);
    throw error;
  }
};

/**
 * Get session by ID
 * @param {string} sessionId - Session ID
 * @returns {Object|null} Session
 */
export const getSession = async (sessionId) => {
  try {
    const session = await ProjectorSession.findOne({ sessionId })
      .populate('classId', 'grade section classCode')
      .populate('startedBy', 'name email')
      .populate('institutionId', 'name');

    return session;
  } catch (error) {
    logger.error('Get session error:', error);
    throw error;
  }
};

/**
 * Update session content
 * @param {string} sessionId - Session ID
 * @param {Object} contentData - Content data
 * @returns {Object} Updated session
 */
export const updateContent = async (sessionId, contentData) => {
  try {
    const session = await ProjectorSession.findOne({ sessionId, status: 'active' });
    
    if (!session) {
      throw new Error('Session not found or not active');
    }

    await session.updateContent(contentData);
    
    logger.info(`Session ${sessionId} content updated`);
    return session;
  } catch (error) {
    logger.error('Update content error:', error);
    throw error;
  }
};

/**
 * Connect device to session
 * @param {string} sessionId - Session ID
 * @param {string} deviceId - Device ID
 * @param {string} deviceName - Device name
 * @returns {Object} Updated session
 */
export const connectDevice = async (sessionId, deviceId, deviceName) => {
  try {
    const session = await ProjectorSession.findOne({ sessionId, status: 'active' });
    
    if (!session) {
      throw new Error('Session not found or not active');
    }

    await session.addDevice(deviceId, deviceName);
    
    logger.info(`Device ${deviceId} connected to session ${sessionId}`);
    return session;
  } catch (error) {
    logger.error('Connect device error:', error);
    throw error;
  }
};

/**
 * End session
 * @param {string} sessionId - Session ID
 * @returns {Object} Ended session
 */
export const endSession = async (sessionId) => {
  try {
    const session = await ProjectorSession.findOne({ sessionId });
    
    if (!session) {
      throw new Error('Session not found');
    }

    session.status = 'ended';
    await session.save();
    
    logger.info(`Session ${sessionId} ended`);
    return session;
  } catch (error) {
    logger.error('End session error:', error);
    throw error;
  }
};

File: mobile/lib/features/projector/services/projector_service.dart
import '../../../core/services/api_service.dart';

/// Projector Service
/// Phase 2.5: K-12 Multi-Access
class ProjectorService {
  final ApiService _apiService;

  ProjectorService({ApiService? apiService})
      : _apiService = apiService ?? ApiService();

  /// Create projector session
  Future<Map<String, dynamic>> createSession({
    required String deviceId,
    required String institutionId,
    String? classId,
    Map<String, dynamic>? currentContent,
  }) async {
    try {
      final response = await _apiService.post(
        '/projector/sessions',
        data: {
          'deviceId': deviceId,
          'institutionId': institutionId,
          if (classId != null) 'classId': classId,
          if (currentContent != null) 'currentContent': currentContent,
        },
      );

      final data = response.data as Map<String, dynamic>;
      return data['data'] as Map<String, dynamic>? ?? {};
    } catch (e) {
      print('❌ Create session error: $e');
      rethrow;
    }
  }

  /// Get active session for device
  Future<Map<String, dynamic>?> getActiveSession(String deviceId) async {
    try {
      final response = await _apiService.get('/projector/sessions/device/$deviceId');
      final data = response.data as Map<String, dynamic>;
      if (data['success'] == true && data['data'] != null) {
        return data['data'] as Map<String, dynamic>;
      }
      return null;
    } catch (e) {
      print('❌ Get active session error: $e');
      return null;
    }
  }

  /// Get session by ID
  Future<Map<String, dynamic>> getSession(String sessionId) async {
    try {
      final response = await _apiService.get('/projector/sessions/$sessionId');
      final data = response.data as Map<String, dynamic>;
      return data['data'] as Map<String, dynamic>? ?? {};
    } catch (e) {
      print('❌ Get session error: $e');
      rethrow;
    }
  }

  /// Update session content
  Future<Map<String, dynamic>> updateContent(
    String sessionId,
    Map<String, dynamic> contentData,
  ) async {
    try {
      final response = await _apiService.put(
        '/projector/sessions/$sessionId/content',
        data: {
          'contentData': contentData,
        },
      );

      final data = response.data as Map<String, dynamic>;
      return data['data'] as Map<String, dynamic>? ?? {};
    } catch (e) {
      print('❌ Update content error: $e');
      rethrow;
    }
  }

  /// Connect device to session
  Future<Map<String, dynamic>> connectDevice(
    String sessionId,
    String deviceId,
    String deviceName,
  ) async {
    try {
      final response = await _apiService.post(
        '/projector/sessions/$sessionId/connect',
        data: {
          'deviceId': deviceId,
          'deviceName': deviceName,
        },
      );

      final data = response.data as Map<String, dynamic>;
      return data['data'] as Map<String, dynamic>? ?? {};
    } catch (e) {
      print('❌ Connect device error: $e');
      rethrow;
    }
  }

  /// End session
  Future<void> endSession(String sessionId) async {
    try {
      await _apiService.post('/projector/sessions/$sessionId/end');
    } catch (e) {
      print('❌ End session error: $e');
      rethrow;
    }
  }
}

